### PR TITLE
Add AMP for validation, prediction and testing

### LINKED ISSUE #6566: Native mixed precision OOM Val Step
## 🐛 Bug

`NativeMixedPrecisionPlugin` only implements `train_step_context` which can lead to Out of Memory Errors during validation/testing .

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

Performed a second check with the Apex Backend (Level O1) and does not run out of memory.

### Expected behavior
Should not run OOM during validation/testing..

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### Additional context
@justusschock

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds AMP Autocast for validation, prediction and testing

Fixes #6566

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Awesome catch ! Thanks a lot !

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
@justusschock this is a great catch, any chance you could update the changelog/milestones?

**Review by SeanNaren — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: tests/models/test_amp.py [hunk 1]
```diff
@@ -23,19 +23,34 @@
...
         return {"loss": loss}
 
+    def training_step(self, batch, batch_idx):
+        return self._step(batch, batch_idx)
+
+    def validation_step(self, batch, batch_idx):
+        self._step(batch, batch_idx)
+
+    def test_step(self, batch, batch_idx):
+        self._step(batch, batch_idx)
```
Comment (awaelchli) [inline tests/models/test_amp.py:46]: You are missing the return statements
Boring model has epoch_end methods implemented that want the outputs

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_amp.py (hunk 1 above)
Comment (awaelchli) [reply tests/models/test_amp.py:46]: this will probably fix the tests

Comment (awaelchli) [reply tests/models/test_amp.py:46]:
```suggestion
    def validation_step(self, batch, batch_idx):
        return self._step(batch, batch_idx)

    def test_step(self, batch, batch_idx):
        return self._step(batch, batch_idx)
```

Comment (awaelchli) [reply tests/models/test_amp.py:47]: The epoch_end methods require it.
You can also do `model.validation_epoch_end = None` I suppose
In any case, in order to make the tests pass something has to change

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@justusschock mind check the failing GPU test?

File: tests/models/test_amp.py (hunk 1 above)
Comment (awaelchli) [reply tests/models/test_amp.py]: my intuition was wrong, adding the return is not enough. The training step returns a key "loss" but the validation epoch end hooks try to access a key "x" returned from validation_step.

**Pull request merged**